### ISSUE TITLE
Bearer Token Issue when using an AAD and an MSA

### ISSUE THREAD

**Issue by LanceMcCarthy** [LanceMcCarthy/MvpApi#41]
For some profile photos, the Base64 is invalid or is failing to convert:

```
Exception Format_BadBase64Char. 
```

**Comment by LanceMcCarthy** [comment]
Attempting to apply a MIME check before decoding the base 64 might resolve this. Since the first few characters of the string contain the file extension, I'll automatically have the MIME type!

The following should do the trick:

```
base64String = base64String.TrimStart('"').TrimEnd('"');

var data = base64String.Substring(0, 5);

var fileExtension = string.Empty;

switch (data.ToUpper())
{
    case "IVBOR":
        fileExtension = "png";
        break;
    case "/9J/4":
        fileExtension = "jpg";
         break;
}

var imageData = base64String.Substring(5, base64String.Length);
var imgBytes = Convert.FromBase64String(imageData);
```

**Comment by ReedKimble** [comment]
Tried the update from a local build, got this error message now:
The input is not a valid Base-64 string as it contains a non-base 64 character, more than two padding characters, or an illegal character among the padding characters.

**Comment by LanceMcCarthy** [comment]
Thanks @ReedKimble this brings me closer to the problem with some images. It seems that there are still characters. 

Since you're running it locally, can you put a breakpoint here 

https://github.com/LanceMcCarthy/MvpApi/blob/e5235e289bb15b8e7dbd7e5efe8ec9bb35348707/MvpApi.Uwp/Services/MvpApiService.cs#L180

and copy the value of `base64string` into your clipboard and email it to me? This would help me fix this once and for all.

**Comment by ReedKimble** [comment]
Well, that is interesting…. Check this out, this is the value of base64string:

{"Message":"Not able to find user","MessageDetail":"The e-mail address given, reedkimble@live.com, did not match to an ID for user lookup","StatusCode":400}

I’ll note that when I originally became an MVP it was under a different email address than I use now.  I had two linked to my live.com profile, the old one being a company email.  I’m no longer with that company so just use the live.com address now.  The old email was rkimble@microplastics.com<mailto:rkimble@microplastics.com> though I’m not sure that helps.

I wonder if this is something that Betsy will have to sort out for us on the MVP site side of things?

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

**Comment by LanceMcCarthy** [comment]
AHA! That, I believe, is the common thread between the folks who have this issue.

I'm in the middle of reaching out to some folks internally (and in the WinDev alias) to find out who works on the API. I'll update this thread as soon as I know more.

This looks like a legitimate API bug.

**Comment by ReedKimble** [comment]
Excellent!  Glad we found something to go on.  I'll keep an eye out for an update and test again when needed.

**Comment by LanceMcCarthy** [comment]
Small update. Definitely a common thread. I'm in the process of testing using multiple account login (Azure AD and MSA). This is a completely different approach using UWP's `WebAccountProvider` (that square popup you form the OS that has your already logged in accounts). 

I've already got MSA working in a test branch, but Azure AD requires additional tokens for access. I'll let you know when you can try the test branch.

**Comment by LanceMcCarthy** [comment]
@ReedKimble  This may have been fixed in #52 , during testing for a different OAuth issue @Depechie discovered that his Base64 started working after re-adding the scopes.

Let me know after you test v1.6.1 when it rolls out. Thanks!

**Comment by LanceMcCarthy** [comment]
@ReedKimble Can you try the latest update (1.7) and tell me if the problem is resolved?

**Comment by LanceMcCarthy** [comment]
Close as duplicate thread confirmed fix. Will reopen if Reed replies with information that problem still exists for him.

**Comment by ReedKimble** [comment]
I’m trying to update but keep getting an error (Code: 0x80073CFB) during the download.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

**Comment by LanceMcCarthy** [comment]
Hi @ReedKimble , 

I've seen this before across a multitude of apps, it's a windows Store problem :/

One thing that comes to mind, is that the new version has a minimum requirement of Fall Creators Update (Windows 10, v 16299), are you running an earlier version of Windows 10?

A few things you can try:

1. Uninstall the current version first.
2. If that doesn't work, uninstall again and manually delete the app's local installation folder:

**C:\Users\[USERNAME]\AppData\Local\Packages\61469LanceLotSoftware.MVPCompanion_m6f3850p49bnt\**

3. Use the Windows Store repair tool, this will fix store caching issues (it doesn't touch app data)
    1. Open start menu
    2. Search for `wsreset` (you'll need to type the whole thing before it shows int he results)
    3. Run the command

**Comment by ReedKimble** [comment]
I think my OS version is ok – 1803 build 17134.285

I did an uninstall and wsreset then was able to install the 1.7 version.

Unfortunately, the error with the Base64 still exists…

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

**Comment by LanceMcCarthy** [comment]
It is expected to see the Base64 error upon first launch of the app after the update for many user who have a stored token. 

However, close it and then relaunch the app again. On the second launch, you should be automatically logged in.

**Comment by ReedKimble** [comment]
It does automatically log in – even after resinstalling the login was still automatic.

But no matter how many times I launch the program, I still get the Base64 error.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

**Comment by LanceMcCarthy** [comment]
Thanks Reed. I'm going to redesign my error catching logic so that it shows the proper error message. The Base64 message isn't technically correct, the issue is you're getting a 404.

thank you for your patience and cooperation with testing the updates!

**Comment by LanceMcCarthy** [comment]
@ReedKimble Can you give today's update a try and let me know if you're able to sign in this time? https://www.microsoft.com/store/productid/9NRXNX3WLH77

**Comment by ReedKimble** [comment]
Here’s what happened:

Just after waking my PC and seeing your email, I got the windows notification that the app had updated.  I launched it and was presented with the standard Microsoft login experience.  I entered my email and password and the app opened.  There were no errors or messages.

However, as you can see in the screenshots attached, everything is blank.  I don’t see any contribution info and my profile page is empty.

So that’s where we are at as of today’s update.  If there’s anything else you’d like me to try, please let me know.

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

**Comment by adamj7** [comment]
I also experienced this authentication issue. I was able to solve it by changing my “Primary Alias” setting to the non MSA email address.

https://account.microsoft.com/profile/ (or https://account.microsoft.com and click on your name in the top left corner)
Click "Manage how you sign in to Microsoft" beside your email address.
If you don't already have it, add your MVP associated email address to the list
Make your MVP associated email the Primary account Alias.

![MVPCompanion-Workaround](https://user-images.githubusercontent.com/18252822/75261851-ab233900-57b9-11ea-8890-343e42779348.png)

Load the MVP Companion app again and it should work.

**Comment by LanceMcCarthy** [comment]
Closing this is as unsolvable via client-side app, you can find a solution if you follow @adamj7's guidance in this reply https://github.com/LanceMcCarthy/MvpApi/issues/41#issuecomment-590924727 

Note
If you are getting an error after logging in, this was likely due to having more contributions in your MVP history than can be transmitted in a single HTTP response. This has been fixed in 2022.213.1 along with the home page revamp, you can install it now using the beta channel at https://dvlup.blob.core.windows.net/general-app-files/Installers/MvpCompanion/index.html
